File: backend/app/schemas/graph.py
"""知识图谱 Pydantic Schema"""

from typing import Optional, List
from uuid import UUID
from pydantic import BaseModel, Field


class GraphNodeItem(BaseModel):
    id: UUID
    name: str
    entity_type: str
    group_id: int
    weight: int
    properties: Optional[dict] = None

    model_config = {"from_attributes": True}


class GraphEdgeItem(BaseModel):
    id: UUID
    source_entity_id: UUID
    target_entity_id: UUID
    source_name: Optional[str] = None
    target_name: Optional[str] = None
    relation_type: str
    relation_desc: Optional[str] = None
    weight: float = 1.0

    model_config = {"from_attributes": True}


class GraphSubgraphResponse(BaseModel):
    nodes: List[GraphNodeItem] = []
    edges: List[GraphEdgeItem] = []


# ── 实体抽取请求 / 响应 ──


class GraphExtractRequest(BaseModel):
    """实体抽取请求"""
    text: str = Field(..., min_length=1, max_length=50000, description="待抽取文本")
    source_doc_id: Optional[UUID] = Field(None, description="来源文档 ID（可选，用于溯源）")


class ExtractedTripleItem(BaseModel):
    """单条抽取结果"""
    source: str
    target: str
    relation: str
    source_type: str = ""
    target_type: str = ""


class GraphExtractResponse(BaseModel):
    """实体抽取响应"""
    triples: List[ExtractedTripleItem] = []
    nodes_created: int = 0
    edges_created: int = 0

File: api/graph.ts
/**
 * 知识图谱 API
 */
import { api } from "./client";

export interface GraphNode {
  id: string;
  name: string;
  entity_type: string;
  weight: number;
  source_doc_id?: string;
  created_at: string;
}

export interface GraphEdge {
  id: string;
  source_entity_id: string;
  target_entity_id: string;
  relation_type: string;
  relation_desc?: string;
  source_name: string;
  target_name: string;
}

export async function apiGetGraphNodes(filters?: {
  entity_type?: string;
  keyword?: string;
}) {
  const params: Record<string, string> = {};
  if (filters?.entity_type) params.entity_type = filters.entity_type;
  if (filters?.keyword) params.keyword = filters.keyword;

  const res = await api.get<GraphNode[]>("/graph/nodes", params);
  return res.data;
}

export async function apiGetGraphEdges() {
  const res = await api.get<GraphEdge[]>("/graph/edges");
  return res.data;
}

export async function apiGetSubgraph(centerNode: string, depth = 2) {
  const res = await api.get<{ nodes: GraphNode[]; edges: GraphEdge[] }>(
    "/graph/subgraph",
    { center_node: centerNode, depth: String(depth) },
  );
  return res.data;
}

export async function apiSearchGraphNodes(q: string, limit = 20) {
  const res = await api.get<GraphNode[]>("/graph/search", {
    q,
    limit: String(limit),
  });
  return res.data;
}

export async function apiExtractEntities(text: string, sourceDocId?: string) {
  const res = await api.post<{
    triples: {
      source: string;
      target: string;
      relation: string;
      source_type: string;
      target_type: string;
    }[];
    nodes_created: number;
    edges_created: number;
  }>("/graph/extract", { text, source_doc_id: sourceDocId });
  return res.data;
}

/* ── 节点 CRUD ── */

export async function apiUpdateGraphNode(
  nodeId: string,
  updates: {
    name?: string;
    entity_type?: string;
    weight?: number;
    properties?: Record<string, unknown>;
  },
) {
  const res = await api.put<GraphNode>(`/graph/nodes/${nodeId}`, updates);
  return res.data;
}

export async function apiDeleteGraphNode(nodeId: string) {
  const res = await api.delete<{ deleted: boolean }>(`/graph/nodes/${nodeId}`);
  return res.data;
}

export async function apiDeleteGraphNodes(ids: string[]) {
  const res = await api.post<{ deleted: number }>("/graph/nodes/batch-delete", {
    ids,
  });
  return res.data;
}
